### PR TITLE
Fix boolean literal types

### LINKED ISSUE #10432: Bug: Boolean type literals misbehaving
**TypeScript Version:**  nightly (2.1.0-dev.20160818)

**Code**
I lifted this from Anders's final example in https://github.com/Microsoft/TypeScript/pull/9407#issue-162831239.
```ts
type Result<T> = { success: true, value: T } | { success: false };

function foo(): Result<number> {
    if (someTest()) {
        return { success: true, value: 42 };
    }
    else {
        return { success: false };
    }
}
function someTest(): Boolean { return true; }
```

**Expected behavior:**
Type-checks OK

**Actual behavior:** 
Type error:
```
result.ts(5,16): error TS2322: Type '{ success: boolean; value: number; }' is not assignable to type 'Result<number>'.
  Type '{ success: boolean; value: number; }' is not assignable to type '{ success: false; }'.
    Types of property 'success' are incompatible.
      Type 'boolean' is not assignable to type 'false'.
result.ts(8,16): error TS2322: Type '{ success: boolean; }' is not assignable to type 'Result<number>'.
  Type '{ success: boolean; }' is not assignable to type '{ success: false; }'.
    Types of property 'success' are incompatible.
      Type 'boolean' is not assignable to type 'false'.
```

If I add manual type annotations (`true as true`, `false as false`) to the `success` assignments it works:
```ts
function foo(): Result<number> {
    if (someTest()) {
        return { success: true as true, value: 42 };
    }
    else {
        return { success: false as false };
    }
}
```

The bug only affects Boolean literals. If I use enum, number, or string type literals it works - for example:
```ts
type Result<T> = { success: 1, value: T } | { success: 0 };

function foo(): Result<number> {
    if (someTest()) {
        return { success: 1, value: 42 };
    }
    else {
        return { success: 0 };
    }
}
```

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #10432.

This PR introduces two changes:

* When an expression has the contextual type `boolean` we now consider it a literal type context. Previously we specifically excluded this case.
* When a binding element has a default value, we now give it a union type of the destructured type and the default type. Previously we'd give it only the destructured type and require the default type to be assignable to the destructured type.

Some background on the changes.

Since `boolean` is *identical* to the union type `true | false`, when an expression has the contextual type `boolean` we really should consider it a literal type context (because the contextual type is a union of unit types). Before this PR we specifically excluded that case in the name of backwards compatibility. With the fix we no longer do. However, that causes another issue to appear:

```ts
let { x = true } = { x: cond ? false : undefined };
```

In the example above, the expression `cond ? false : undefined` is contextually typed by type `boolean` because of the default value on the left hand side. That causes us to consider the expression a literal type context, and we therefore use type `false` for the `false` literal. So, the type of `x` on the right hand side becomes `false | undefined`. Now, when we destructure `x` we require the default value to be of a type assignable to `false | undefined`. But `boolean` isn't.

The issue really is that when a binding element has a default value, we ought to give it a union type of the destructuring type and the default type. With this PR we do. The downside is that you might sometimes end up with a union type when an error might have been more appropriate.

@mhegazy @DanielRosenwasser You might want to take a look.

### CODE REVIEW

File: tests/baselines/reference/contextuallyTypedBindingInitializerNegative.types [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+interface StringUnion {
+>StringUnion : StringUnion
+
+    prop: "foo" | "bar";
+>prop : "foo" | "bar"
+}
+function h({ prop = "baz" }: StringUnion) {}
+>h : ({prop}: StringUnion) => void
+>prop : "foo" | "bar" | "baz"
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/contextuallyTypedBindingInitializerNegative.types]: The initializer is completely ignoring the type annotation in this case. I would consider this a bug.

**Comment by ahejlsberg** [comment]
@mhegazy Want to take a look? I think we need this one in 2.0.

**Comment by mhegazy** [comment]
👍

**Comment by mhegazy** [comment]
merged in release-2.0 in https://github.com/Microsoft/TypeScript/commit/2cd4d20f7bbdf7a9a476ed98a2e5c9f409666419

**Pull request merged**